### PR TITLE
feat: allow different host through CLI flag

### LINKED ISSUE #919: Use 0.0.0.0 for livereload address
## 🚀 Feature

At the moment live reload is hardcoded to use localhost for its address which stops it working inside a docker container. This request is to change it to 0.0.0.0 to allow livereload to work when visited from the host machine.

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Motivation

To enable Docusaurus to be containerised it would be handy to have livereload work from outside the container.

## Pitch

If a developer is able to view the output of livereload served from within a container then a Docusaurus instance can be packaged in a node based Docker container and a self contained environment can be given to other developers so they can write documentation without having node installed locally.

### PR DESCRIPTION
**Pull request by palmer-cl**
## Motivation

This would allow folks running Docusaurus in a container to access the site on localhost.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan
From: https://github.com/facebook/Docusaurus/pull/920 which appears closed.

- Install a fresh Docuaurus in the host environment.
- Run `yarn start` in the host environment.
- Ensure the site is served at http://localhost:3000 as usual.
- Containerize the application (get it installed in a node based container)
- Run `yarn start` within the container.
- Confirm the site is not accessible at http://localhost:3000
- Stop the webserver (live reload).
- Apply the change from this merge request commit.
- Run `yarn start` within the container.
- Confirm the site is accessible at http://localhost:3000 

Note: I have tested this both locally and running inside a Docker container.

## Related PRs
https://github.com/facebook/Docusaurus/pull/920

Fixes #919 
(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js [hunk 1]
```diff
@@ -20,7 +20,10 @@ describe('get reload script', () => {
   test('when server started, returns url with correct port', () => {
     const port = 1234;
     liveReloadServer.start(port);
-    const expectedUrl = `http://localhost:${port}/livereload.js`;
+    const expectedUrl =
+      process.platform === 'win32'
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js]: On macOS, `process.platform` is `'darwin'` and the host I'll get is 0.0.0.0, which probably wouldn't work.

File: packages/docusaurus-1.x/lib/server/liveReloadServer.js [hunk 2]
```diff
@@ -29,7 +29,9 @@ function start(port) {
 
 const getReloadScriptUrl = () => {
   const port = process.env.LIVERELOAD_PORT;
-  return `http://localhost:${port}/livereload.js`;
+  const server = process.platform === 'win32' ? 'localhost' : '0.0.0.0';
```
Comment (yangshun) [inline packages/docusaurus-1.x/lib/server/liveReloadServer.js]: Nit - It's more of `host` than a `server`. Same issue regarding OS as above.

**Comment by palmer-cl** [comment]
@yangshun Great feedback. I can make some changes.

File: packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js (hunk 1 above)
Comment (palmer-cl) [reply packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js]: @yangshun Which do you think would be better? 

Could account for all options on [this list](https://nodejs.org/api/process.html#process_process_platform) or pass it in with a [cli flag](https://docusaurus.io/docs/en/commands#using-arguments) or [env variable](https://docusaurus.io/docs/en/commands#docusaurus-start).

File: packages/docusaurus-1.x/lib/server/liveReloadServer.js (hunk 2 above)
Comment (palmer-cl) [reply packages/docusaurus-1.x/lib/server/liveReloadServer.js]: Ill make this adjustment with a commit alongside the change for https://github.com/facebook/Docusaurus/pull/1475#discussion_r284958730

File: packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js (hunk 1 above)
Comment (yangshun) [reply packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js]: I think CLI would be better. cc @endiliey

Comment (palmer-cl) [reply packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js]: Talked with @endiliey. Will implement CLI flag once this is fixed: https://github.com/facebook/Docusaurus/issues/1427#issuecomment-493341612

Comment (endiliey) [reply packages/docusaurus-1.x/lib/server/__tests__/liveReloadServer.test.js]: Make it like this
https://github.com/facebook/Docusaurus/blob/89ef4b9c440b596077249b8a69ac413187e0cac5/packages/docusaurus/bin/docusaurus.js#L82

**Comment by palmer-cl** [comment]
@endiliey Let me know what you think of this implementation.

File: website-1.x/versioned_docs/version-1.8.0/api-commands.md [hunk 3]
```diff
@@ -147,6 +147,7 @@ This command will build the static website, apply translations if necessary, and
 | Options           | Default | Description                                                                                                                          |
 | ----------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------ |
 | `--port <number>` | `3000`  | The website will be served from port 3000 by default, but if the port is taken up, Docusaurus will attempt to find an available one. |
+|`--host <number>`|`localhost`|Use a custom host with localhost as the default.|
```
Comment (endiliey) [inline website-1.x/versioned_docs/version-1.8.0/api-commands.md]: why is it host <number> ?

Comment (palmer-cl) [reply website-1.x/versioned_docs/version-1.8.0/api-commands.md]: Per https://github.com/facebook/Docusaurus/pull/1475/files/326d299e5baeaee3f7dc6791eb96bf05de82767a#discussion_r284958920

Comment (endiliey) [reply website-1.x/versioned_docs/version-1.8.0/api-commands.md]: Oops my bad. I mean why `host <number>`.

GitHub swallowed the `<number>`

image

Comment (palmer-cl) [reply website-1.x/versioned_docs/version-1.8.0/api-commands.md]: Ha. All good. Ya, I figured number as in 127.0.0.1 but maybe 'value' is better?

Comment (endiliey) [reply website-1.x/versioned_docs/version-1.8.0/api-commands.md]: Follow v2

https://github.com/facebook/Docusaurus/blob/382b88bb49cc13adb920dbb0b901afa44921b041/packages/docusaurus/bin/docusaurus.js#L82

**Pull request merged**